### PR TITLE
[Feature Store] Support time range on ingested jobs

### PR DESCRIPTION
**Pull request by katyakats**

### CODE REVIEW

File: mlrun/feature_store/api.py [hunk 1]
```diff
@@ -261,12 +265,36 @@ def ingest(
             raise mlrun.errors.MLRunInvalidArgumentError(
                 "data source was not specified"
             )
+
+        if source.schedule:
+            featureset.reload(update_spec=False)
+            if featureset.status.targets and featureset.status.targets[0].last_written:
```
Comment (dinal) [inline mlrun/feature_store/api.py:271]: last_written will be set only on parquet targets? 
if so, then if the first target is nosql and the second in parquet you won't go into the if statement at all

File: mlrun/feature_store/api.py [hunk 2]
```diff
@@ -261,12 +265,36 @@ def ingest(
...
+                min_time = datetime.fromisoformat(
+                    featureset.status.targets[0].last_written
+                )
+                for target in featureset.status.targets:
+                    if target.last_written:
+                        cur_last_written = datetime.fromisoformat(target.last_written)
+                        if cur_last_written < min_time:
+                            min_time = cur_last_written
+
+                source.start_time = min_time
```
Comment (dinal) [inline mlrun/feature_store/api.py:282]: first time ingest won't have any last_written set? then min should be 0?

File: mlrun/feature_store/api.py [hunk 3]
```diff
@@ -261,12 +265,36 @@ def ingest(
...
+        if source.schedule:
+            featureset.reload(update_spec=False)
+            if featureset.status.targets and featureset.status.targets[0].last_written:
+
+                min_time = datetime.fromisoformat(
+                    featureset.status.targets[0].last_written
+                )
+                for target in featureset.status.targets:
+                    if target.last_written:
+                        cur_last_written = datetime.fromisoformat(target.last_written)
```
Comment (dinal) [inline mlrun/feature_store/api.py:278]: i don't have a specific comment but something is worrying me about time zones conversion here (and in datetime.now)
need to think it through

File: mlrun/feature_store/api.py [hunk 4]
```diff
@@ -261,12 +265,32 @@ def ingest(
             raise mlrun.errors.MLRunInvalidArgumentError(
                 "data source was not specified"
             )
+
+        if source.schedule:
+            featureset.reload(update_spec=False)
+            min_time = datetime.date.max
+            for target in featureset.status.targets:
+                if target.last_written:
+                    cur_last_written = datetime.fromisoformat(target.last_written)
```
Comment (dinal) [inline mlrun/feature_store/api.py:274]: if last_written has time zone convert it to same (same in datetime.now())

File: mlrun/feature_store/feature_set.py [hunk 5]
```diff
@@ -207,6 +208,13 @@ def targets(self, targets: List[DataTarget]):
     def update_target(self, target: DataTarget):
         self._targets.update(target)
 
+    def update_last_written_for_target(
+        self, target_path: str, last_written: datetime.datetime
+    ):
+        for target in self._targets:
+            if target.path == target_path:
+                target.last_written = last_written
+
```
Comment (dinal) [inline mlrun/feature_store/feature_set.py:217]: need to update the status? or sync to db somehow?

File: tests/system/feature_store/test_feature_store.py [hunk 6]
```diff
@@ -831,6 +832,84 @@ def test_unaggregated_columns(self):
         }
     )
 
+    def test_schedule_on_filtered_by_time(self):
```
Comment (dinal) [inline tests/system/feature_store/test_feature_store.py:835]: some more tests permutations - 
1. add data in the past (before last_written) and see it's not read
2. use several targets in different order
3. verify partitioned parquet

File: mlrun/feature_store/feature_set.py (hunk 5 above)
Comment (katyakats) [reply mlrun/feature_store/feature_set.py:219]: sync to db is done inside ingest

File: mlrun/feature_store/api.py [hunk 7]
```diff
@@ -261,12 +265,35 @@ def ingest(
             raise mlrun.errors.MLRunInvalidArgumentError(
                 "data source was not specified"
             )
+
+        if source.schedule:
+            featureset.reload(update_spec=False)
+            max_datetime = datetime(date.max.year, date.max.month, date.max.day)
```
Comment (dinal) [inline mlrun/feature_store/api.py:271]: there's datetime.max

File: mlrun/feature_store/api.py [hunk 8]
```diff
@@ -261,12 +265,35 @@ def ingest(
...
+            for target in featureset.status.targets:
+                if target.last_written:
+                    cur_last_written = datetime.fromisoformat(target.last_written)
+                    if cur_last_written < min_time:
+                        min_time = cur_last_written
+            if min_time != max_datetime:
+                source.start_time = min_time
+                time_zone = min_time.tzinfo
+                current_time = datetime.now()
+                source.end_time = current_time.replace(tzinfo=time_zone)
```
Comment (dinal) [inline mlrun/feature_store/api.py:282]: datetime.now() can receive tz

Comment (dinal) [reply mlrun/feature_store/api.py:282]: shouldn't source.end_time be set in any case?

File: tests/system/feature_store/test_feature_store.py [hunk 9]
```diff
@@ -832,6 +833,114 @@ def test_unaggregated_columns(self):
...
+                    now + pd.Timedelta(minutes=10),
+                    pd.Timestamp("2021-01-09 13:00:00"),
+                ],
+                "first_name": ["moshe", "dina", "katya", "uri"],
+                "data": [50, 10, 25, 30],
+            }
+        )
+        # writing down a remote source
+        target2 = ParquetTarget()
+        data_set = fs.FeatureSet("data", entities=[Entity("first_name")])
```
Comment (dinal) [inline tests/system/feature_store/test_feature_store.py:918]: you don't need to redefine target2 and data_set, they are already defined above

**Review by dinal — APPROVED** [review]

**Review by Hedingber — APPROVED** [review]

**Pull request merged**